js/render/draw_symbol.js

```diff
@@ -33,165 +31,181 @@ function drawSymbols(painter, source, layer, coords) {
     painter.depthMask(false);
     gl.disable(gl.DEPTH_TEST);
 
-    var tile, elementGroups, bucket;
+    drawLayerSymbols(painter, source, layer, coords, false,
+            layer.paint['icon-translate'],
+            layer.paint['icon-translate-anchor'],
+            layer.layout['icon-rotation-alignment'],
+            layer.layout['icon-size'],
+            layer.paint['icon-halo-width'],
+            layer.paint['icon-halo-color'],
+            layer.paint['icon-halo-blur'],
+            layer.paint['icon-opacity'],
+            layer.paint['icon-color']);
+
+    drawLayerSymbols(painter, source, layer, coords, true,
+            layer.paint['text-translate'],
+            layer.paint['text-translate-anchor'],
+            layer.layout['text-rotation-alignment'],
+            layer.layout['text-size'],
+            layer.paint['text-halo-width'],
+            layer.paint['text-halo-color'],
+            layer.paint['text-halo-blur'],
+            layer.paint['text-opacity'],
+            layer.paint['text-color']);
 
-    for (var i = 0; i < coords.length; i++) {
-        tile = source.getTile(coords[i]);
-        bucket = tile.getBucket(layer);
-        if (!bucket) continue;
-        elementGroups = bucket.elementGroups;
-        if (!elementGroups.icon.length) continue;
+    gl.enable(gl.DEPTH_TEST);
 
-        painter.enableTileClippingMask(coords[i]);
-        drawSymbol(painter, layer, coords[i].posMatrix, tile, bucket, elementGroups.icon, 'icon', elementGroups.sdfIcons, elementGroups.iconsNeedLinear);
-    }
+    drawCollisionDebug(painter, source, layer, coords);
+}
+
+function drawLayerSymbols(painter, source, layer, coords, isText,
+        translate,
+        translateAnchor,
+        rotationAlignment,
+        size,
+        haloWidth,
+        haloColor,
+        haloBlur,
+        opacity,
+        color) {
```

Comment:
I agree that the large number of parameters is awkward but I prefer this to an options object because:
- this lists what parameters the function accepts. With an options object you need either document it separately or look where the function is called.
- none of these parameters are optional. Putting them in an options object is more likely to hide typos since they'll just be `undefined` instead of jslint errors. On the other hand this approach could suffer from ordering bugs.


I prefer keeping this as is